gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientTest.java

```diff
@@ -0,0 +1,111 @@
+package com.google.cloud.hadoop.gcsio;
+
+import static com.google.cloud.hadoop.gcsio.MockGoogleCloudStorageImplFactory.mockedGcsClientImpl;
+import static com.google.cloud.hadoop.util.testing.MockHttpTransportHelper.mockTransport;
+import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.assertEquals;
+import static org.mockito.AdditionalAnswers.answersWithDelay;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import com.google.api.client.testing.http.MockHttpTransport;
+import com.google.api.core.ApiFutures;
+import com.google.cloud.storage.Storage;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.IntFunction;
+import org.junit.Test;
+
+public class GoogleCloudStorageClientTest {
+
+  private static final String TEST_BUCKET_NAME = "foo-bucket";
+
+  private static final String TEST_OBJECT_NAME = "foo-object";
+
+  private static MockHttpTransport transport = mockTransport();
+
+  @Test
+  public void readVectored_successfulRead()
+      throws IOException, ExecutionException, InterruptedException, TimeoutException,
+          URISyntaxException {
+    GoogleCloudStorageClientImpl gcsClientImpl = getMockedGcsClientImpl(false);
+    IntFunction<ByteBuffer> allocator = (length) -> ByteBuffer.allocateDirect(length);
+    List<VectoredIORange> ranges = getListOfVectoredIORange();
+
+    VectoredIOMetrics result =
+        gcsClientImpl.readVectored(
+            ranges, allocator, new URI("gs://" + TEST_BUCKET_NAME + "/" + TEST_OBJECT_NAME));
+
+    assertEquals(getReadVectoredData(ranges.get(0)), FakeBlobReadSession.SUBSTRING_20_10);
+    assertEquals(getReadVectoredData(ranges.get(1)), FakeBlobReadSession.SUBSTRING_50_7);
+    assertEquals(getReadVectoredData(ranges.get(2)), FakeBlobReadSession.SUBSTRING_65_17);
+
+    assertThat(result.getClientInitializationDuration()).isLessThan(100L);
+    assertThat(result.getReadDuration()).isLessThan(100L);
```

Comment:
If these times are important to your test -- or any tests for that matter -- I recommend providing a clock or function to the instance under test where you can control the advancement of the client in your test instead of relying on wall time to cooperate.

Reply:
Makes sense. Need some broader changes rather than simply mocking the SUT, will do in future cls.